### ISSUE TITLE
Keep players from posting picks before commish is ready

### ISSUE THREAD

**Issue by blreams** [blreams/collegefootballpick10#54]
We have multiple commissioners that co-manage the pool. Each week one of the co-commissioners is assigned as the CoW (Commissioner-O'-the-Week). That commissioner is responsible for coming up with the slate of games for the week. This is usually done in a collaborative way with the other commissioners providing input.

Imagine this scenario. CoW creates a slate of games on Sunday and wants to get feedback from the other commissioners. He enters the games using createweek URL and waits. In the meantime, we don't want any players making picks until the slate is solidified.

What about just setting Week.lock_picks = datetime.now() which would effectively keep players from posting picks (you are checking the pick deadline before posting picks...right?)? Since I am only holding the LOCK during the execution of the view's post() method, it is a simple matter to get the current value for Week.lock_picks and squirrel it away. Once the post() method is ready to return (either because the edits have been saved to the DB, or because there was a form error) it can restore Week.lock_picks.

**Comment by jbholden** [comment]
I'm a little confused.  Are you describing 1 or 2 scenarios?

* scenario 1:  set lock picks while collaborating on games   (long duration)
* scenario 2:  set lock picks during the post() method  (short duration) (which page is this?)

My initial thoughts are that this shouldn't be a problem.
I think there might be a side effect though on the pool state and I might need to change some of the messages to better reflect the state.  Another issue might be displaying the pick deadline.

**Better messages**
If the pick deadline has expired, then the enter picks page shows a message saying the deadline has passed.  A better message might be to say that the picks are locked, to avoid confusion when setting lock_picks to some early date

**Pick deadline**
I don't know if somewhere on the site you want to display the pick deadline.  If this is the case, then when setting lock_picks to an early date, the pick deadline will not be displayed with the correct time.  

You could do something like if the lock_picks date is greater than 1 week in the past, then picks are locked and the pick deadline is invalid.  If the deadline is within 1 week, then lock_picks represents the pick deadline.

Another possibility might be to change lock_picks to boolean and add a pick_deadline to the Pick model that is a datetime.

**Pool State**
pick10/database.py get_pool_state() function
This function will behave a little bit differently, but I believe it is not a problem.

We could add another state like 'setup_games' that happens before 'enter_picks'.  We would need to determine what criteria to check for this state.

```Python
List of Pool States
# these 1st 2 states represent beginning and end of the pool
'not_started'    #  pool not started, year has week 1 model, but no games yet
'end_of_year'  #  week 13 is FINAL

# these states get cycled thru each week
'enter_picks'  # week exists but before pick deadline
'week_not_started'  # last week in NOT_STARTED state
'week_in_progress' # last week in IN_PROGRESS state
'week_final'  #  if not week 13, last week in FINAL state
```

Scenario 1:  Start of pool
The current sequence goes from 'not_started' to 'enter_picks' to 'week_not_started'
If you lock picks early, the sequence will be 'not_started' to 'week_not_started' to 'enter_picks' to 'week_not_started'

Scenario 2:  Start of a week (say week 2)
The current sequence goes from 'week_final' to 'enter_picks' to 'week_not_started'
If you lock picks early, the sequence will be 'week_final' to 'week_not_started' to 'enter_picks' to 'week_not_started

**Comment by blreams** [comment]
I think I did confuse this issue a little when I was thinking about Issue #48. But I think the idea about how to lock picks is similar.

I like your suggestion about Better Messages. Essentially picks are locked if the value of lock_picks is in the past.

Displaying the pick deadline as a time delta is only important if picks are not locked (IOW if the value of lock_picks is in the past). Otherwise, it is OK to simply state that the deadline has passed, or picks are locked, or whatever.

As for Pool State...what you are really talking about is Pool State (not started, in progress, end of year) and then a separate Week State for each Week (week not started, enter picks, week in progress, week final). We might want to open a separate Issue for this. In the past we have had the situation where one game in Week N was delayed such that Week N was not final until after Week N+1 was in the enter picks phase. In that situation, what would be the state of the Pool (from your definitions above)? I think it makes more sense to track the state of each Week independently. 

Do you remember what happened the week of 9/11/2001? All the games on the following Saturday were called off. Most were replayed later in the year. We've had to change games because of volcanic eruptions and hurricanes.

**Comment by jbholden** [comment]
For reference, the pool state came from this issue:  https://github.com/jbholden/cdcpool_google/issues/20

I think if we make major changes to the pool state and week state at this point might require rewriting and retesting a bunch of code.  The code I have written uses "not_started", "in_progress", and "final" for week states extensively.  Adding a "enter_picks" state will require reanalyzing the code.

While your description does seem to make sense for the pool state and week state, I think for now we should just go with it as is and make changes as a possible enhancement in the future or in response to specific problems we find.

I think we can agree on the following
* use the lock_picks variable to lock picks by setting old dates
* updating the pages to show better messages
* only show pick deadlines when picks are not locked

**Comment by blreams** [comment]
Sounds like a plan.

**Comment by blreams** [comment]
Closing this issue.
